### PR TITLE
Add utility for opening remote files with ``fsspec``

### PR DESCRIPTION
**Pull request by jrbourbeau**
- [x] Closes https://github.com/pydata/xarray/issues/9723
- [x] Closes https://github.com/pydata/xarray/issues/8423
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

Supersedes https://github.com/pydata/xarray/pull/8424

### CODE REVIEW

File: xarray/backends/common.py [hunk 1]
```diff
@@ -178,6 +178,15 @@ def _normalize_path_list(
     return _normalize_path_list(paths)
 
 
+def _open_file(file, mode, storage_options=None):
+    import fsspec
+
+    fs, _, paths = fsspec.get_fs_token_paths(
+        file, mode=mode, storage_options=storage_options
+    )
+    return fs.open(paths[0], mode=mode)
```
Comment (jrbourbeau) [inline xarray/backends/common.py:187]: @dcherian I think this get's at your comment here https://github.com/pydata/xarray/issues/9723#issuecomment-2460464611 about restricting to a simple utility function using `fsspec`, but let me know if I'm missing something

File: xarray/backends/common.py [hunk 2]
```diff
@@ -178,6 +178,15 @@ def _normalize_path_list(
     return _normalize_path_list(paths)
 
 
+def _open_file(file, mode, storage_options=None):
```
Comment (jhamman) [inline xarray/backends/common.py:181]:
```suggestion
def _open_remote_file(file, mode, storage_options=None):
```

File: xarray/backends/h5netcdf_.py [hunk 3]
```diff
@@ -149,9 +150,14 @@ def open(
         decode_vlen_strings=True,
         driver=None,
         driver_kwds=None,
+        storage_options=None,
     ):
         import h5netcdf
 
+        if isinstance(filename, str) and is_remote_uri(filename) and driver is None:
+            mode_ = "rb" if mode == "r" else mode
+            filename = _open_file(filename, mode=mode_, storage_options=storage_options)
```
Comment (jhamman) [inline xarray/backends/h5netcdf_.py:159]:
```suggestion
            filename = _open_remote_file(filename, mode=mode_, storage_options=storage_options)
```

**Review by jhamman — APPROVED** [review]
This looks good @jrbourbeau. Thanks for separating the open logic into a nice little utility function.

**Comment by jrbourbeau** [comment]
Yeah, no problem, just added an entry

File: xarray/backends/h5netcdf_.py [hunk 4]
```diff
@@ -149,9 +150,16 @@ def open(
         decode_vlen_strings=True,
         driver=None,
         driver_kwds=None,
+        storage_options=None,
```
Comment (headtr1ck) [inline xarray/backends/h5netcdf_.py:153]: Could we get type hints at least here?

**Review by headtr1ck — APPROVED** [review]

File: xarray/backends/h5netcdf_.py (hunk 4 above)
Comment (jrbourbeau) [reply xarray/backends/h5netcdf_.py]: Yeah, just pushed a commit for that 👍

Comment (jrbourbeau) [reply xarray/backends/h5netcdf_.py]: ...and fixed things to make `mypy` happy (one day that'll happen on the first try)

**Comment by jrbourbeau** [comment]
Thanks all

**Pull request merged**